libyul/backends/evm/AsmCodeGen.cpp

```diff
@@ -146,22 +146,32 @@ pair<shared_ptr<AbstractAssembly>, AbstractAssembly::SubID> EthAssemblyAdapter::
 	return {make_shared<EthAssemblyAdapter>(*assembly), static_cast<size_t>(sub.data())};
 }
 
-void EthAssemblyAdapter::appendDataOffset(AbstractAssembly::SubID _sub)
+void EthAssemblyAdapter::appendDataOffset(std::vector<AbstractAssembly::SubID> const& _subIds)
 {
-	auto it = m_dataHashBySubId.find(_sub);
-	if (it == m_dataHashBySubId.end())
-		m_assembly.pushSubroutineOffset(_sub);
-	else
-		m_assembly << evmasm::AssemblyItem(evmasm::PushData, it->second);
+	yulAssert(_subIds.size() < 32, "");
+	if (_subIds.size() == 1)
+		if (auto it = m_dataHashBySubId.find(_subIds[0]); it != m_dataHashBySubId.end())
+		{
+			m_assembly << evmasm::AssemblyItem(evmasm::PushData, it->second);
+			return;
+		}
+
+	u256 encodedSubId = evmasm::Assembly::encodeSubIds(_subIds);
+	m_assembly.pushSubroutineOffset(encodedSubId);
 }
 
-void EthAssemblyAdapter::appendDataSize(AbstractAssembly::SubID _sub)
+void EthAssemblyAdapter::appendDataSize(std::vector<AbstractAssembly::SubID> const& _subIds)
 {
-	auto it = m_dataHashBySubId.find(_sub);
-	if (it == m_dataHashBySubId.end())
-		m_assembly.pushSubroutineSize(static_cast<size_t>(_sub));
-	else
-		m_assembly << u256(m_assembly.data(h256(it->second)).size());
+	yulAssert(_subIds.size() < 32, "");
```

Comment:
I don't think this is enforced anywhere in the analyzer. If a user tries to do more than 31 nesting levels, he'll get an ICE due to a failed assertion. I think we should have a proper error check in addition to the assertions so that the user gets an error message instead.

Could you also add a simple test case with many nesting levels?

BTW, I'd use `<=` to make it more apparent that 31 is the limit, not 32:
```suggestion
	yulAssert(_subIds.size() <= 31, "");
```